### ISSUE TITLE
Possibility to access logs on the Bot Framework Platform ?

### ISSUE THREAD

**Issue by jbdrucker** [Microsoft/BotBuilder#2374]
_TL;DR : Out bot stopped working with no apparent reason and I'd like to check wether the messages arrive on Microsoft Bot Framework servers_

Hello all,

Our bot stopped working one morning without any apparent reason. We checked everything, but we failed to find "why". We even went as far as to comment everything and just make the bot echo a default message like this :

    bot.use({
        botbuilder : function(session,next) {
            session.send("Work in progress, sorry !");
           // next();
    
        },  
        send : function(event,next) {  
            console.log(event);  
            next();  
        }  
    });  

Note the commented `next` function, no further code from any dialog is executed except the send part.

So far :
- works fine in the emulator when hosted locally
- works fine in the emulator when hosted remotely
- we can see in the logs the messages arriving from the webchat and any other set channel
- we can see in the logs the sent message ( `console.log(event)` )
- while doing a tcpdump capture on our proxy, we see the TLS traffic towards the different platforms (directline.botframework.com and so on) (but can't see the content of the encrypted traffic)

**BUT** no message appears on the webchat, nor in any other channel (skype or facebook).

We're honestly at a loss. We spent a whole day trying everything, from downgrading versions to even create a new bot with new credentials but same endpoint : fail.

So I'm wondering : is there any way to check wether the botframework platform actually receives our messages, is there an error log, can we inspect the content of the messages etc.. in short : tooling on the remote end ?

Thanks in advance for all your ideas and leads

**Comment by mcclaskiem** [comment]
I think this is somewhat similar to the issue I am seeing #2372

**Comment by sumitdobhal** [comment]
I got the following error when I published my bot application in server and acessed it remotely.
It's  saying to Configure ngrok which I have configured in the emulator.
Still I am facing the error.
If anyone knows this error please reply.

 **Error: The bot is remote, but the callback URL is localhost. Without tunneling software you will not receive replies. 
Fix it: Configure ngrok 
Learn more: Connecting to bots hosted remotely 
-> POST Refresh access token failed with status code: 400**

**Comment by jbdrucker** [comment]
@sumitdobhal It's a totally different problem. You have to configure ngrok separately from the emulator, like registering your ngrok account, setting your proxy information and things like that.

Please don't hijack my already complicated issue ;)

**Comment by Stevenic** [comment]
Adding Dan as it sounds like a DirectLine issue.

**Comment by dandriscoll** [comment]
Hi @jbdrucker, we log errors in three ways:

1. If your bot sends an invalid message, we return an HTTP error. You should trap/log this.
2. If we try to call your bot and it fails, we add an "Issue" to your bot in the Bot Framework developer portal. Open your bot's dashboard and look at at the "Issues" column for each channel.
3. We log detailed error and tracing information to your Application Insights (telemetry) instance, if you have it configured. Simply set it up and paste the Instrumentation Key into your bot's dashboard on dev.botframework.com.

Since the problem is occurring across multiple channels, I expect it has to do with the bot or its configuration. If you're seeing inbound but not outbound calls, I would start by looking at the HTTP result from those channels. If you get an HTTP 200 back, we received, processed, and delivered the message your bot sent.

**Comment by jbdrucker** [comment]
Sorry for the lack of response, I've been pretty busy (so don't close this yet :p )

Point 2 : No issues are reported on the Bot Framework developer portal.
Point 3 : I'll look into Application Insights, I thought we had to be hosted on Azure for this to work but apparently that's not the case

Point 1 : I'm still looking at how I can do such a thing. Correct me if I'm wrong, but the ChatConnector only manage the incoming messages (listening to events coming from the Bot Connector/Framework), and `request` is used for the outgoing communications (sending requests to facebook/directline/younameit.botframework.com. I'd have to go deep inside the source code in order totrace that, no ?

**Comment by jbdrucker** [comment]
Still investigating. What I newly found :

Setting NODE_DEBUG="request" at startup is giving useful information (I don't know why I haven't thought of that earlier). Here is the blurred-out log :

> ChatConnector: message received.
REQUEST { method: 'GET',
  url: 'https://state.botframework.com/v3/botstate/webchat/users/USERID',
  json: true,
  headers: { 'User-Agent': 'Microsoft-BotFramework/3.1 (BotBuilder Node.js/3.6.0)' },
  callback: [Function] }
REQUEST { method: 'GET',
  url: 'https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID',
  json: true,
  headers: { 'User-Agent': 'Microsoft-BotFramework/3.1 (BotBuilder Node.js/3.6.0)' },
  callback: [Function] }
REQUEST make request https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST make request https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST onRequestResponse https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:26 GMT',
  connection: 'close' }
REQUEST reading response's body
REQUEST finish init function https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST response end https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:26 GMT',
  connection: 'close' }
REQUEST end event https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST invalid JSON received https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST emitting complete https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST { method: 'GET',
  url: 'https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID',
  json: true,
  headers: { 'User-Agent': 'Microsoft-BotFramework/3.1 (BotBuilder Node.js/3.6.0)' },
  callback: [Function] }
REQUEST make request https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST onRequestResponse https://state.botframework.com/v3/botstate/webchat/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:27 GMT',
  connection: 'close' }
REQUEST reading response's body
REQUEST finish init function https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST response end https://state.botframework.com/v3/botstate/webchat/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:27 GMT',
  connection: 'close' }
REQUEST end event https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST invalid JSON received https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST emitting complete https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST { method: 'GET',
  url: 'https://state.botframework.com/v3/botstate/webchat/users/USERID',
  json: true,
  headers: { 'User-Agent': 'Microsoft-BotFramework/3.1 (BotBuilder Node.js/3.6.0)' },
  callback: [Function] }
REQUEST make request https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST onRequestResponse https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:28 GMT',
  connection: 'close' }
REQUEST reading response's body
REQUEST finish init function https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST response end https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:28 GMT',
  connection: 'close' }
REQUEST end event https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST invalid JSON received https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST emitting complete https://state.botframework.com/v3/botstate/webchat/conversations/CONVOID/users/USERID
REQUEST onRequestResponse https://state.botframework.com/v3/botstate/webchat/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:28 GMT',
  connection: 'close' }
REQUEST reading response's body
REQUEST finish init function https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST response end https://state.botframework.com/v3/botstate/webchat/users/USERID 403 { 'cache-control': 'no-cache',
  pragma: 'no-cache',
  'content-length': '0',
  expires: '-1',
  server: 'Microsoft-IIS/8.0',
  'x-powered-by': 'ASP.NET',
  'x-content-type-options': 'nosniff',
  date: 'Mon, 20 Mar 2017 15:57:28 GMT',
  connection: 'close' }
REQUEST end event https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST invalid JSON received https://state.botframework.com/v3/botstate/webchat/users/USERID
REQUEST emitting complete https://state.botframework.com/v3/botstate/webchat/users/USERID

As you may notice, the call to state.botframework.com responds a 403 error, stating it received an invalid JSON message. But that would mean the framework is doing something wrong since it's not a call I'm making but a kind of internal call.

I tried to set `persistUserData` and `persistConversationData ` to `false` and restart the bot to see if those calls were still made, but no change happened.

But I have no way of knowing what's this JSON made of yet, and

**Comment by nwhitmont** [comment]
@jbdrucker are you still experiencing the issue?

**Comment by d1mnewz** [comment]
@nwhitmont hope he doens't, but actually I do. 
I faced this problem with bot in Slack. However, in WebChat/localhost it works properly. 
Already restarted App Service, double-checked credentials on slack config on https://dev.botframework.com/bots , created totally new SlackApp with new credentials, reverted commits on Azure. 
and the same endpoint: zero response from bot in Slack. 
Where can I find some logs? Does my bot even gets my messages from Slack? Dashboard shows 0 errors. What else can I do? 

Config: VS2017, C#, Azure App Service, continuous deployment from Github, Entity Framework, SQL Server on Azure, SQL Database on Azure.

**Comment by mcclaskiem** [comment]
@d1mnewz Not sure if this will help but I had a similar issue and it ended up being that on the server the env file was not being read but it is built in a way that that fails silently and didnt show errors anywhere. I had to add a node-module that read the env file and everything began working as intended.

**Comment by d1mnewz** [comment]
@mcclaskiem 
> env file
you mean environment variables that are set on Azure App Service? Application Settings?

**Comment by mcclaskiem** [comment]
If you are following the tutorial to get this set up they don't include a `require` for the env file since technically you need to authenticate when working locally. This then ends up failing silently on the server and without proper auth the bot will not be able to respond back to your selected messaging service. I am not sure how Azure handles environment settings so as long as you can confirm that the Microsoft APP_ID and KEY are being read by the application the server then you can rule this out.

**Comment by d1mnewz** [comment]
I don't follow any tutorial, actually I don't think I have any `require` in my code.
But I got your message, are you talking about `[BotAuthentication]` attribute? In C# we put this on top of class that handles messages. We already checked that one :)

Anyway, thank you 👍

**Comment by nwhitmont** [comment]
@jbdrucker reopen if you're still having issues. thx

**Comment by jbdrucker** [comment]
Well I still have the problem, but I kind of gave up. It seems like the problem is either out of my reach or so small (like a accented character messing something up) it would require to re-do everything from scratch.

We created an App Insights account and linked it to the bot (through the portal), it gives very little information (the only event being "Message received from X to X"). We can only notice that no messages are received by the framework from our endpoint, and the log I posted above shows why : the request fails because the JSON the framework sends is somehow messed up.

So we moved on, we rebuilt that specific bot without the Bot Framework, and sadly we're keeping in mind that for future projects the framework could hang up on us at anytime without warning.

I know our architecture is not the targeted one for Microsoft (read "not hosted on Azure") so it's also not helping, in a way, however I think that's the solution most professional actors will choose anyway. 

So far setting the framework in debug mode helps one with debugging the "dialog part" (you see each step and so on), but very little is shown regarding the "environment debug" ("Sending message" yeah sure why not, but that's it, no return code or anything AFAICS).

So here we are ;)

**Comment by neonithinr** [comment]
@jbdrucker  Im facing the exact same problem. Just read all the comments. So what have you ultimately decided. Sorry if the question is too naive :)

**Comment by neonithinr** [comment]
@jbdrucker I meant to ask what you mean by rebuilt bot "without BOT Framework" . A little more info on that please.

**Comment by jbdrucker** [comment]
By "without the bot framework" I meant exactly that : our bot is targeted for Facebook Messenger, so we use the FB SDK to build our bots without using the BotBuilder library. Like you would do if the bot framework didn't exist.
